Objective-C/Tests/URLEndpointListenerTest.m

```diff
@@ -893,4 +893,114 @@ - (void) testCloseWithActiveListener {
+    Assert(replicator.serverCertificate == NULL);
+    
+    [replicator start];
+    
+    [self waitForExpectations: @[x1] timeout: timeout];
+    Assert(replicator.serverCertificate == NULL);
+    
+    [replicator stop];
+    
+    [self waitForExpectations: @[x2] timeout: timeout];
+    Assert(replicator.serverCertificate == NULL);
+    
+    [self stopListen];
+}
+
+- (void) testReplicatorServerCertWithTLS {
+    if (!self.keyChainAccessAllowed) return;
+    
+    XCTestExpectation* x1 = [self expectationWithDescription: @"idle"];
+    XCTestExpectation* x2 = [self expectationWithDescription: @"stopped"];
+    
+    Listener* listener = [self listen];
+    
+    SecCertificateRef serverCert = (__bridge SecCertificateRef) listener.tlsIdentity.certs[0];
+    CBLReplicator* replicator = [self replicator: self.otherDB
+                                       continous: YES
+                                          target: listener.localEndpoint
+                                      serverCert: serverCert];
+    [replicator addChangeListener: ^(CBLReplicatorChange *change) {
+        CBLReplicatorActivityLevel activity = change.status.activity;
+        if (activity == kCBLReplicatorIdle)
+            [x1 fulfill];
+        else if (activity == kCBLReplicatorStopped && !change.status.error)
+            [x2 fulfill];
+    }];
+    Assert(replicator.serverCertificate == NULL);
+    
+    [replicator start];
+    
+    [self waitForExpectations: @[x1] timeout: timeout];
+    Assert(replicator.serverCertificate != NULL);
+    [self checkEqualForCert: serverCert andCert: replicator.serverCertificate];
+    
+    [replicator stop];
+    
+    [self waitForExpectations: @[x2] timeout: timeout];
+    Assert(replicator.serverCertificate != NULL);
+    [self checkEqualForCert: serverCert andCert: replicator.serverCertificate];
+    
+    [self stopListen];
+}
+
+- (void) testReplicatorServerCertWithTLSError {
+    if (!self.keyChainAccessAllowed) return;
+    
+    XCTestExpectation* x1 = [self expectationWithDescription: @"stopped"];
+    
+    Listener* listener = [self listen];
+    
+    CBLReplicator* replicator = [self replicator: self.otherDB
+                                       continous: YES
+                                          target: listener.localEndpoint
+                                      serverCert: nil];
+    [replicator addChangeListener: ^(CBLReplicatorChange *change) {
+        CBLReplicatorActivityLevel level = change.status.activity;
+        if (level == kCBLReplicatorStopped && change.status.error)
+            [x1 fulfill];
+    }];
+    Assert(replicator.serverCertificate == NULL);
+    
+    [replicator start];
+    
+    [self waitForExpectations: @[x1] timeout: timeout];
+    Assert(replicator.serverCertificate != NULL);
+    
+    SecCertificateRef serverCert = (__bridge SecCertificateRef) listener.tlsIdentity.certs[0];
+    [self checkEqualForCert: serverCert andCert: replicator.serverCertificate];
+    
```

Comment:
make sure, same replicator can connect after the failure?

Reply:
Done